### PR TITLE
Fix in product sync while fixing unresolved references.

### PR DESCRIPTION
**Pull request by ahmetoz**
#580 

- Tests
    - [X] Unit 
    - [X] Integration
- [X] Documentation
- [x] Add Release Notes entry.

### CODE REVIEW

File: src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java [hunk 1]
```diff
@@ -386,7 +386,7 @@ public ProductReferenceResolver(@Nonnull final ProductSyncOptions productSyncOpt
         final List<CompletionStage<Map<String, String>>> futures = new ArrayList<>();
 
         final Set<String> productKeys = referencedKeys.getProductKeys();
-        if (!productKeys.isEmpty()) {
+        if (productKeys.isEmpty()) {
```
Comment (JudeNiroshan) [inline src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:389]: why execute the reference fetch when we do not have any keys?

Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java]: no reason, it's typo, reverted: https://github.com/commercetools/commercetools-sync-java/pull/581/commits/5c22820a06e5c339a127aa7a6a5b6d8f30ee7f60

File: src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java [hunk 2]
```diff
@@ -432,6 +432,6 @@ public ProductReferenceResolver(@Nonnull final ProductSyncOptions productSyncOpt
         }
 
         return collectionOfFuturesToFutureOfCollection(futures, toList())
-            .thenApply(maps -> productKeys.isEmpty() ? Collections.emptyMap() : maps.get(0));
+            .thenCompose(ignored -> productService.cacheKeysToIds(Collections.emptySet())); // return all cache.
```
Comment (JudeNiroshan) [inline src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:435]: I guess it's the time to introduce a getter for the caches in the service layer. wdyt?

Comment (JudeNiroshan) [inline src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:435]: also naming is confusing. `productService.cacheKeysToIds` when the user ask nothing, it shouldn't return everything it has.`getKeysToIdsCache` would be good candidate name for the new method

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsInAnyOrderIT.java [hunk 3]
```diff
@@ -112,6 +112,7 @@ private ProductSyncOptions buildSyncOptions() {
                                         .errorCallback((syncException, draft, product, updateActions)
                                             -> collectErrors(syncException.getMessage(), syncException))
                                         .beforeUpdateCallback(this::collectActions)
+                                        .batchSize(1)
```
Comment (JudeNiroshan) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsInAnyOrderIT.java:115]: why we need to make batch size to 1? how does it relate to productWith500ProductReferences test case?

**Review by JudeNiroshan — CHANGES_REQUESTED** [review]

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsInAnyOrderIT.java (hunk 3 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithReferencedProductsInAnyOrderIT.java:115]: caches are based on the batch, so it's ensuring the keyToId is considering other batches too, before it was not.

File: src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java (hunk 2 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:435]: For now, the cache is an internal part of the service, for this PR, I will focus the part of the quickly fixing the issues with returning cache. We will improve the caching with different PR soon.

Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:435]: Same here, maybe I could add those points to a separate GitHub issue, soon we will describe the new concept for the cache implementation,  https://github.com/commercetools/commercetools-sync-java/pull/581#discussion_r496616879

**Review by JudeNiroshan — APPROVED** [review]

**Review by lojzatran — APPROVED** [review]

File: src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java (hunk 2 above)
Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:435]: Added as a context: https://github.com/commercetools/commercetools-sync-java/issues/582

Comment (ahmetoz) [reply src/main/java/com/commercetools/sync/products/helpers/ProductReferenceResolver.java:435]: Added as a context: https://github.com/commercetools/commercetools-sync-java/issues/582

**Pull request merged**